<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Single cell transcriptomics - Enrichment analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../assets/SIB_logo.svg" rel="icon" type="image/svg+xml">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Source+Sans+Pro">
<!-- Matomo --><script>
  var _paq = window._paq = window._paq || [];
  /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
  _paq.push(['trackPageView']);
  _paq.push(['enableLinkTracking']);
  (function() {
    var u="https://matomo.sib.swiss/";
    _paq.push(['setTrackerUrl', u+'matomo.php']);
    _paq.push(['setSiteId', '220']);
    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
    g.async=true; g.src=u+'matomo.js'; s.parentNode.insertBefore(g,s);
  })();
</script><!-- End Matomo Code -->
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Single cell transcriptomics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../precourse_preparations.html" rel="" target="">
 <span class="menu-text">Precourse preparations</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../course_schedule.html" rel="" target="">
 <span class="menu-text">Course schedule</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-1" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">day 1</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-1">
<li>
    <a class="dropdown-item" href="../day1/day1-1_setup.html" rel="" target="">
 <span class="dropdown-text">Setup</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day1/day1-1a_cellranger.html" rel="" target="">
 <span class="dropdown-text">Introduction &amp; cellranger</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day1/day1-2_analysis_tools_qc.html" rel="" target="">
 <span class="dropdown-text">Analysis tools and QC</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-2" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">day 2</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-2">
<li>
    <a class="dropdown-item" href="../day2/day2-1_normalization_scaling.html" rel="" target="">
 <span class="dropdown-text">Normalization and scaling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day2/day2-2_dimensionality_reduction.html" rel="" target="">
 <span class="dropdown-text">Dimensionality reduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day2/day2-3_integration.html" rel="" target="">
 <span class="dropdown-text">Integration</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day2/day2-4_clustering.html" rel="" target="">
 <span class="dropdown-text">Clustering</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-3" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">day 3</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-3">
<li>
    <a class="dropdown-item" href="../day3/day3-1_cell_annotation.html" rel="" target="">
 <span class="dropdown-text">Cell annotation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day3/day3-2_differential_gene_expression.html" rel="" target="">
 <span class="dropdown-text">Differential gene expression</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day3/day3-3_enrichment_analysis.html" rel="" target="">
 <span class="dropdown-text">Enrichment analysis</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-projects" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">projects</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-projects">
<li>
    <a class="dropdown-item" href="../projects/project1.html" rel="" target="">
 <span class="dropdown-text">Project 1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../projects/project2.html" rel="" target="">
 <span class="dropdown-text">Project 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../projects/project3.html" rel="" target="">
 <span class="dropdown-text">Project 3</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../projects/project4.html" rel="" target="">
 <span class="dropdown-text">Project 4</span></a>
  </li>  
    </ul>
</li>
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/sib-swiss/single-cell-r-training/" rel="" target=""><i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Enrichment analysis</li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../index.html" class="sidebar-logo-link">
      <img src="../assets/SIB_LogoQ_GBv.svg" alt="" class="sidebar-logo py-0 d-lg-inline d-none"></a>
      </div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#material" id="toc-material" class="nav-link active" data-scroll-target="#material">Material</a></li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises">Exercises</a></li>
  <li><a href="#about-orgdb" id="toc-about-orgdb" class="nav-link" data-scroll-target="#about-orgdb">About OrgDb</a></li>
  <li><a href="#clear-environment" id="toc-clear-environment" class="nav-link" data-scroll-target="#clear-environment">Clear environment</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Enrichment analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><section id="material" class="level2"><h2 class="anchored" data-anchor-id="material">Material</h2>
<ul>
<li><a href="http://www.gsea-msigdb.org/gsea/msigdb/index.jsp">MSigDB</a></li>
<li>
<code>clusterProfiler</code> <a href="https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html">vignette</a>
</li>
<li><a href="http://revigo.irb.hr/">Revigo</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/SPIA.html">Signaling Pathway Impact Analysis (SPIA)</a></li>
<li>Original <a href="https://www.pnas.org/content/102/43/15545">paper</a> on GSEA</li>
<li>
<a href="https://string-db.org/">STRING</a> for protein-protein interactions</li>
<li>
<a href="https://gitlab.com/evogenlab/GO-Figure">GO figure!</a> for plotting GO terms and the associated <a href="https://www.frontiersin.org/articles/10.3389/fbinf.2021.638255/full">paper</a>
</li>
</ul></section><section id="exercises" class="level2"><h2 class="anchored" data-anchor-id="exercises">Exercises</h2>
<p>Load the following packages:</p>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-2_5eecd515b9acb22b0513ab55ec4cadfe">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/contribution-knowledge-mining/">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/">enrichplot</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If the <code>FindMarkers</code> or <code>FindAllMarkers</code> functions were used, we obtained a table listing only the significant genes, but we don’t have any information of fold change for the non-significant genes. Therefore, we can use the over-representation analysis which is a threshold-based method. Using our list of significant genes, we can test if any gene set is over-represented among significant genes or not using a test similar to a Fisher test to compare differences in proportions.</p>
<p>The <code>clusterProfiler</code> package provides functions for over-representation analysis of Gene Ontology gene sets (among other functions, including functions for actual GSEA) or KEGG gene sets.</p>
<p>Genes can be labeled using different types of labels, eg symbol, Ensembl ID, Entrez ID. To list the allowed label types use:</p>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-3_15b46edc084b8998912db84eb5a5f95f">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"org.Hs.eg.db"</span>, update <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">keytypes</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="about-orgdb" class="level2 callout-info"><h2 class="anchored" data-anchor-id="about-orgdb">About OrgDb</h2>
<p>For other organisms, you can find available OrgDbs at <a href="http://bioconductor.org/packages/release/BiocViews.html#___OrgDb">bioconductor</a></p>
</section><p>Let’s select a set of genes that are downregulated in the tumor cells compared to normal:</p>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-4_38dcc90263d3159de58fdcc0b4b0f093">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tum_down</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">limma_de</span>,</span>
<span>                    <span class="va">limma_de</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fl">1</span> </span>
<span>                      <span class="op">&amp;</span> <span class="va">limma_de</span><span class="op">$</span><span class="va">adj.P.Val</span> <span class="op">&lt;</span>  <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">tum_down_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">tum_down</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can do a Gene Ontology term over-representation analysis based on this set of genes. Make sure you check out the help of this function to understand its arguments:</p>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-5_88ab584e492a19b95c750d27fc4b1dcd">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="op">?</span><span class="va">enrichGO</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-6_5eaa4d16987534171924ef8da59211b4">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tum_vs_norm_go</span> <span class="op">&lt;-</span> <span class="fu">clusterProfiler</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html">enrichGO</a></span><span class="op">(</span><span class="va">tum_down_genes</span>,</span>
<span>                                            <span class="st">"org.Hs.eg.db"</span>,</span>
<span>                                            keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                                            ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>                                            minGSSize <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The results are stored in the <code>@result</code> slot:</p>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-7_9f1ce66395e30d85e42070052ec93fa3">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="va">tum_vs_norm_go</span><span class="op">@</span><span class="va">result</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-8_df681c58e0f581fe3756f34101ae9ba8">
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 10%">
<col style="width: 10%">
<col style="width: 36%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 6%">
<col style="width: 8%">
<col style="width: 6%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">ID</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">GeneRatio</th>
<th style="text-align: left;">BgRatio</th>
<th style="text-align: right;">pvalue</th>
<th style="text-align: right;">p.adjust</th>
<th style="text-align: right;">qvalue</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">GO:0007059</td>
<td style="text-align: left;">GO:0007059</td>
<td style="text-align: left;">chromosome segregation</td>
<td style="text-align: left;">105/809</td>
<td style="text-align: left;">424/18870</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0098813</td>
<td style="text-align: left;">GO:0098813</td>
<td style="text-align: left;">nuclear chromosome segregation</td>
<td style="text-align: left;">79/809</td>
<td style="text-align: left;">312/18870</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GO:0000070</td>
<td style="text-align: left;">GO:0000070</td>
<td style="text-align: left;">mitotic sister chromatid segregation</td>
<td style="text-align: left;">62/809</td>
<td style="text-align: left;">184/18870</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0000819</td>
<td style="text-align: left;">GO:0000819</td>
<td style="text-align: left;">sister chromatid segregation</td>
<td style="text-align: left;">67/809</td>
<td style="text-align: left;">225/18870</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GO:0140014</td>
<td style="text-align: left;">GO:0140014</td>
<td style="text-align: left;">mitotic nuclear division</td>
<td style="text-align: left;">71/809</td>
<td style="text-align: left;">274/18870</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0000280</td>
<td style="text-align: left;">GO:0000280</td>
<td style="text-align: left;">nuclear division</td>
<td style="text-align: left;">88/809</td>
<td style="text-align: left;">441/18870</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
The columns <code>GeneRatio</code> and <code>BgRatio</code>
</div>
</div>
<div class="callout-body-container callout-body">
<p>The columns <code>GeneRatio</code> and <code>BgRatio</code> that are in the <code>enrichResult</code> object represent the numbers that are used as input for the Fisher’s exact test.</p>
<p>The two numbers (M/N) in the <code>GeneRatio</code> column are:</p>
<ul>
<li>M: Number of genes of interest (in our case <code>tum_down_genes</code>) that are in the GO set</li>
<li>N: Number of genes of interest with any GO annoation.</li>
</ul>
<p>The two numbers (k/n) in the <code>BgRatio</code> column are:</p>
<ul>
<li>k: Number of genes in the universe that are in the GO set</li>
<li>n: Number of genes in the universe with any GO annoation</li>
</ul>
<p>A low p-value resulting from the Fisher’s exact means that M/N is signficantly greater than k/n.&nbsp;</p>
</div>
</div>
<p>Some GO terms seem redundant because they contain many of the same genes, which is a characteristic of Gene Ontology gene sets. We can simplify this list by removing redundant gene sets:</p>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-9_7d1f941a44638ed949d2fd43662845d3">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">enr_go</span> <span class="op">&lt;-</span> <span class="fu">clusterProfiler</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/simplify-methods.html">simplify</a></span><span class="op">(</span><span class="va">tum_vs_norm_go</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-10_2bb864afaffdd8959914abebb5a83389">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="va">enr_go</span><span class="op">@</span><span class="va">result</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-11_de113d5378a46d8b0447571ff56e2c2c">
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 10%">
<col style="width: 10%">
<col style="width: 36%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 6%">
<col style="width: 8%">
<col style="width: 6%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">ID</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">GeneRatio</th>
<th style="text-align: left;">BgRatio</th>
<th style="text-align: right;">pvalue</th>
<th style="text-align: right;">p.adjust</th>
<th style="text-align: right;">qvalue</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">GO:0007059</td>
<td style="text-align: left;">GO:0007059</td>
<td style="text-align: left;">chromosome segregation</td>
<td style="text-align: left;">105/809</td>
<td style="text-align: left;">424/18870</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0098813</td>
<td style="text-align: left;">GO:0098813</td>
<td style="text-align: left;">nuclear chromosome segregation</td>
<td style="text-align: left;">79/809</td>
<td style="text-align: left;">312/18870</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GO:0000070</td>
<td style="text-align: left;">GO:0000070</td>
<td style="text-align: left;">mitotic sister chromatid segregation</td>
<td style="text-align: left;">62/809</td>
<td style="text-align: left;">184/18870</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0000280</td>
<td style="text-align: left;">GO:0000280</td>
<td style="text-align: left;">nuclear division</td>
<td style="text-align: left;">88/809</td>
<td style="text-align: left;">441/18870</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GO:0044772</td>
<td style="text-align: left;">GO:0044772</td>
<td style="text-align: left;">mitotic cell cycle phase transition</td>
<td style="text-align: left;">77/809</td>
<td style="text-align: left;">470/18870</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0051983</td>
<td style="text-align: left;">GO:0051983</td>
<td style="text-align: left;">regulation of chromosome segregation</td>
<td style="text-align: left;">40/809</td>
<td style="text-align: left;">131/18870</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We can quite easily generate a plot called an enrichment map with the <code>enrichplot</code> package:</p>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-12_d5ac3111d596ed0c277f5c37ca446679">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">enrichplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/emapplot.html">emapplot</a></span><span class="op">(</span><span class="fu">enrichplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/pairwise_termsim.html">pairwise_termsim</a></span><span class="op">(</span><span class="va">enr_go</span><span class="op">)</span>,</span>
<span>                     showCategory <span class="op">=</span> <span class="fl">30</span>,</span>
<span>                     cex.params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>category_label <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="day3-3_enrichment_analysis_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Instead of testing for Gene Ontology terms, we can also test for other gene set collections. For example the Hallmark collection from <a href="http://www.gsea-msigdb.org/gsea/msigdb/index.jsp">MSigDB</a>:</p>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-13_801f0cfa8ae419315026c2f1910a9d7a">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gmt</span> <span class="op">&lt;-</span> <span class="fu">msigdbr</span><span class="fu">::</span><span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr.html">msigdbr</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"human"</span>, category <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can use the function <code>enricher</code> to test for over-representation of any set of genes of the Hallmark collection. We have to include the “universe”, i.e.&nbsp;the full list of background, non significant genes, against which to test for differences in proportions:</p>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-14_8c2fa0684da77924a2581049576c9aa7">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tum_vs_norm_enrich</span> <span class="op">&lt;-</span> <span class="fu">clusterProfiler</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enricher.html">enricher</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">tum_down_genes</span>,</span>
<span>                                                universe <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">proB</span><span class="op">)</span>,</span>
<span>                                                pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>                                                pvalueCutoff  <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                                                qvalueCutoff  <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                                                TERM2GENE <span class="op">=</span> <span class="va">gmt</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gs_name"</span>, <span class="st">"gene_symbol"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>When using the genes down-regulated in tumor, among the over-represented Hallmark gene sets, we have <code>HALLMARK_G2M_CHECKPOINT</code>, which includes genes involved in the G2/M checkpoint in the progression through the cell division cycle.</p>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-15_4e51908d7e8586574c1fcfb9bfb8b675">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="va">tum_vs_norm_enrich</span><span class="op">@</span><span class="va">result</span><span class="op">[</span><span class="va">tum_vs_norm_enrich</span><span class="op">@</span><span class="va">result</span><span class="op">$</span><span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span>,<span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-16_f2967a1b6f75fb1bda107e59acf7273b">
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 22%">
<col style="width: 22%">
<col style="width: 22%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">ID</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">GeneRatio</th>
<th style="text-align: left;">BgRatio</th>
<th style="text-align: right;">pvalue</th>
<th style="text-align: right;">p.adjust</th>
<th style="text-align: right;">qvalue</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">HALLMARK_E2F_TARGETS</td>
<td style="text-align: left;">HALLMARK_E2F_TARGETS</td>
<td style="text-align: left;">HALLMARK_E2F_TARGETS</td>
<td style="text-align: left;">80/344</td>
<td style="text-align: left;">195/3866</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="even">
<td style="text-align: left;">HALLMARK_G2M_CHECKPOINT</td>
<td style="text-align: left;">HALLMARK_G2M_CHECKPOINT</td>
<td style="text-align: left;">HALLMARK_G2M_CHECKPOINT</td>
<td style="text-align: left;">67/344</td>
<td style="text-align: left;">188/3866</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HALLMARK_MITOTIC_SPINDLE</td>
<td style="text-align: left;">HALLMARK_MITOTIC_SPINDLE</td>
<td style="text-align: left;">HALLMARK_MITOTIC_SPINDLE</td>
<td style="text-align: left;">48/344</td>
<td style="text-align: left;">197/3866</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="even">
<td style="text-align: left;">HALLMARK_MYC_TARGETS_V1</td>
<td style="text-align: left;">HALLMARK_MYC_TARGETS_V1</td>
<td style="text-align: left;">HALLMARK_MYC_TARGETS_V1</td>
<td style="text-align: left;">33/344</td>
<td style="text-align: left;">194/3866</td>
<td style="text-align: right;">0.0001564</td>
<td style="text-align: right;">0.0017991</td>
<td style="text-align: right;">0.0016468</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HALLMARK_ESTROGEN_RESPONSE_LATE</td>
<td style="text-align: left;">HALLMARK_ESTROGEN_RESPONSE_LATE</td>
<td style="text-align: left;">HALLMARK_ESTROGEN_RESPONSE_LATE</td>
<td style="text-align: left;">28/344</td>
<td style="text-align: left;">166/3866</td>
<td style="text-align: right;">0.0005747</td>
<td style="text-align: right;">0.0052875</td>
<td style="text-align: right;">0.0048398</td>
</tr>
</tbody>
</table>
</div>
</div>
<section id="clear-environment" class="level3"><h3 class="anchored" data-anchor-id="clear-environment">Clear environment</h3>
<p>Clear your environment:</p>
<div class="cell" data-hash="day3-3_enrichment_analysis_cache/html/unnamed-chunk-17_acc6a7c790d459f80932dfc519d96746">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/gc.html">gc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">.rs.restartR</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>